### PR TITLE
Feature lp 33 leaderboard api

### PR DESCRIPTION
**Pull request by Charvit123**
## What

Please include a summary of the changes.

- [x] leader board api

# Checklist

- [x] My code follows the style guidelines of this project and is properly formatted
- [x] I have performed a self-review of my own changes
- [x] All existing tests pass, and I have added new ones if required

### CODE REVIEW

File: backend/src/user/user.service.ts [hunk 1]
```diff
@@ -39,4 +69,36 @@ export class UserService {
...
+        },
+      },
+    });
+
+    const users = await Promise.all(
+      userCourse.map(async (currentUserCourse) => {
+        return {
+          user: await this.prismaService.user.findFirst({
+            where: {
+              id: currentUserCourse.userId,
```
Comment (abdulkalam1233) [inline backend/src/user/user.service.ts:95]: use:  user_id: {
      in: users.map(user => user.user_id)
    },
Avoid making DB calls in loops

File: backend/test/user.e2e-spec.ts [hunk 2]
```diff
@@ -45,4 +45,11 @@ describe('UserController (e2e)', () => {
       .set('Authorization', `Bearer ${authToken}`);
     expect(response.status).toBe(200);
   });
+
+  it('user/leaderboard/ (GET) - should return the top 5 users list', async () => {
+    const response = await request(app.getHttpServer())
+      .patch('/user/leaderboard')
+      .set('Authorization', `Bearer ${authToken}`);
+    expect(response.status).toBe(200);
```
Comment (abdulkalam1233) [inline backend/test/user.e2e-spec.ts:53]: Data length check is missing

File: backend/test/user.e2e-spec.ts [hunk 3]
```diff
@@ -45,4 +45,11 @@ describe('UserController (e2e)', () => {
       .set('Authorization', `Bearer ${authToken}`);
     expect(response.status).toBe(200);
```
Comment (abdulkalam1233) [inline backend/test/user.e2e-spec.ts:46]: Add data assertion

File: backend/src/user/user.service.ts [hunk 4]
```diff
@@ -1,18 +1,48 @@
...
     const user = await this.prismaService.user.findFirst({ where: { id } });
     delete user.password;
     return user;
   }
 
-  async getCourseByUserId(userid: string): Promise<Course[]> {
+  async getCourseByUserId(userid: string, status: string): Promise<Course[]> {
+    if (status === 'completed') {
+      const prismaUserCourse = await this.prismaService.userCourse.findMany({
+        where: { userId: userid, isCompleted: true },
```
Comment (abdulkalam1233) [inline backend/src/user/user.service.ts:19]: Make this query dynamic based on the status

File: backend/src/user/user.service.spec.ts
Comment (abdulkalam1233) [inline backend/src/user/user.service.spec.ts:1]: use https://www.npmjs.com/package/jest-mock-extended

Comment (abdulkalam1233) [reply backend/src/user/user.service.spec.ts:1]: const PrismaServiceMockFactory = {
  provide: PrismaService,
  useFactory: mockDeep(PrismaService),
};

**Review by abdulkalam1233 — CHANGES_REQUESTED** [review]

File: backend/src/user/user.service.spec.ts
Comment (Charvit123) [reply backend/src/user/user.service.spec.ts:1]: @abdulkalam1233 we have tried to use this deepmock but we are getting the same errors of circular dependency at three other places.

**Pull request merged**